### PR TITLE
docs: Add documentation

### LINKED ISSUE #424: Improve the Readme
### Summary

The current Readme is way too minimalistic. 

### Details

A proper Readme should at least give an overview of celestia-node and how it relates to tendermint/celestia-core and how 
it augments it. Ideally, it should contain a high-level and easy understandable diagram (additional to links to the ADRs where it makes sense).

For external contributors looking through our repositories it should be very clear what role and which features each 
repository has (including this one). 

It should answer the question: What can you do with it (now and what is intended to be added later)?  

In an ideal world an enumeration with the most relevant packages for external devs would also be contained in the Readme
as well as a link to high-level instructions on how to use node on the command-line.

### PR DESCRIPTION
**Pull request by renaynay**
Adds documentation for: 

- [x] README.md
- [x] header pkg (doc.go)
- [x] shares pkg (doc.go)

Resolves #424 and #426

### CODE REVIEW

**Comment by liamsi** [comment]
Maybe for PRs like this, it could also be interesting to get @YazzyYaz' opinion as well?

File: README.md [hunk 1]
```diff
@@ -1,6 +1,14 @@
 # Celestia Node
```
Comment (liamsi) [inline README.md:1]: Can we add the usual badges here as well?
- godoc 
- code coverage
- lastest release
- etc

Comment (renaynay) [reply README.md:1]: @Bidon15 I added release tag and godoc badges, do you know how to do codecov? I know codecov already has access to celestiaorg IIRC

Comment (liamsi) [reply README.md:1]: Codecov can also happen in a separate PR imo.

Comment (liamsi) [reply README.md:1]: I see that Viet already started: https://github.com/celestiaorg/celestia-node/pull/564

**Review by liamsi — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -1,6 +1,19 @@
...
+
+[![Go Reference](https://pkg.go.dev/badge/github.com/celestiaorg/celestia-node.svg)](https://pkg.go.dev/github.com/celestiaorg/celestia-node)
+![GitHub release (latest by date including pre-releases)](https://img.shields.io/github/v/release/celestiaorg/celestia-node)
```
Comment (liamsi) [inline README.md:4]:
```suggestion
[![GitHub release (latest by date including pre-releases)](https://img.shields.io/github/v/release/celestiaorg/celestia-node)
```

File: README.md [hunk 3]
```diff
@@ -16,6 +27,32 @@ Golang implementation of Celestia's data availability node types (`light` | `ful
...
+```
+```
 make install
 ```
+
+### Node types
+* **Bridge** nodes - bridge nodes relay blocks from the celestia consensus network to the celestia data availability 
+(DA) network
+* **Full** nodes - full nodes fully reconstruct and store blocks by sampling the DA network for shares
+* **Light** nodes - light nodes verify the availability of block data by sampling the DA network for shares
```
Comment (Wondertan) [inline README.md:40]: 👍🏻 

Not sure why do you duplicate "<type> nodes - <type> nodes", e.g. **Bridge** nodes - bridge nodes

File: service/header/doc.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ExtendedHeaders are generated by bridge nodes that listen to the celestia-core network for blocks,
+validate and extend the block data in order to generate the DAH, and request both the ValidatorSet
+and Commit for that block height from the celestia-core network. Bridge nodes then package that data
+up into an ExtendedHeader and broadcast it to the headersub gossipsub topic. All nodes subscribed
+to headersub will receive and validate the ExtendedHeader, and store it, making it available to all
+other dependent services (such as the DataAvailabilitySampler, or DASer) to access.
+
+There are 5 main components in the header package:
+	1. CoreListener listens for new blocks from the celestia-core network (run by bridge nodes only),
+	   extends them, generates a new ExtendedHeader and publishes it to the headersub gossipsub topic.
```
Comment (Wondertan) [inline service/header/doc.go:18]:
```suggestion
	   extends them, generates a new ExtendedHeader, and publishes it to the HeaderSub.
```

File: service/header/doc.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	2. P2PSubscriber listens for new ExtendedHeaders from the celestia data availability (DA) network (via
+	   the headersub gossipsub topic)
```
Comment (Wondertan) [inline service/header/doc.go:20]:
```suggestion
	   the HeaderSub)
```

File: service/header/doc.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+validate and extend the block data in order to generate the DAH, and request both the ValidatorSet
+and Commit for that block height from the celestia-core network. Bridge nodes then package that data
+up into an ExtendedHeader and broadcast it to the headersub gossipsub topic. All nodes subscribed
+to headersub will receive and validate the ExtendedHeader, and store it, making it available to all
+other dependent services (such as the DataAvailabilitySampler, or DASer) to access.
+
+There are 5 main components in the header package:
+	1. CoreListener listens for new blocks from the celestia-core network (run by bridge nodes only),
+	   extends them, generates a new ExtendedHeader and publishes it to the headersub gossipsub topic.
+	2. P2PSubscriber listens for new ExtendedHeaders from the celestia data availability (DA) network (via
```
Comment (Wondertan) [inline service/header/doc.go:19]:
```suggestion
	2. P2PSubscriber listens for new ExtendedHeaders from the Celestia Data Availability (DA) network (via
```

File: service/header/doc.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+	   the headersub gossipsub topic)
+	3. P2PExchange or CoreExchange request ExtendedHeaders from other celestia DA nodes (default for
+	   full and light nodes) or from a celestia-core node connection (bridge nodes only)
+	4. Syncer manages syncing past ExtendedHeaders from either the DA network or a celestia-core
+       connection (bridge nodes only).
```
Comment (Wondertan) [inline service/header/doc.go:24]:
```suggestion
	4. Syncer manages syncing of past and recent ExtendedHeaders from either the DA network or a celestia-core
       connection (bridge nodes only).
```

File: service/header/doc.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+	5. Store manages storing ExtendedHeaders and making them available for access by other dependent services.
+
+For bridge nodes, the general flow of the header Service is as follows:
+	1. CoreListener listens for new blocks from the celestia-core connection
+	2. CoreListener validates the block and generates the ExtendedHeader, simultaneously storing the
+       extended block shares to disk
+	3. CoreListener publishes the new ExtendedHeader to headersub, notifying all subscribed peers of the new
+	   ExtendedHeader
```
Comment (Wondertan) [inline service/header/doc.go:32]:
```suggestion
	3. CoreListener publishes the new ExtendedHeader to HeaderSub, notifying all subscribed peers of the new
	   ExtendedHeader
```

File: service/header/doc.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+	4. Syncer is already subscribed to the headersub topic so it receives new ExtendedHeaders locally from
+       the CoreListener and stores them to disk via Store.
```
Comment (Wondertan) [inline service/header/doc.go:34]:
```suggestion
	4. Syncer is already subscribed to the HeaderSub, so it receives new ExtendedHeaders locally from
       the CoreListener and stores them to disk via Store.
```

File: service/header/doc.go [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+	4a. If the celestia-core connection is started simultaneously with the bridge node, then the celestia-core
+       connection will handle the syncing component, piping every synced block to the CoreListener
+	4b. If the celestia-core connection is already synced to the network, the Syncer handles requesting past
+		headers up to the network head from the celestia-core connection (using CoreExchange rather than P2PExchange).
+
+For full and light nodes, the general flow of the header Service is as follows:
+	1. Syncer listens for new ExtendedHeaders from headersub
```
Comment (Wondertan) [inline service/header/doc.go:41]:
```suggestion
	1. Syncer listens for new ExtendedHeaders from HeaderSub
```

File: service/header/doc.go [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+	2. If there is a gap between the local head of chain and the new, validated network head, the Syncer
+	   kicks off a sync routine to request all ExtendedHeaders between local head and network head.
+	3. While the Syncer is requesting headers between local head and network head, it appends each subsequent
+	   header to the chain via Store as the new local head.
```
Comment (Wondertan) [inline service/header/doc.go:45]:
```suggestion
	3. While the Syncer requests headers between the local head and network head in batches, it appends them to the subjective chain via Store with the last batched header as the new local head.
```

File: service/header/doc.go [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+An ExtendedHeader is similar to a regular block header from celestia-core, but "extended" by adding
+the DataAvailabilityHeader (DAH), ValidatorSet and Commit in order to allow for light and full nodes
+to reconstruct block data by referencing the DAH.
+
+ExtendedHeaders are generated by bridge nodes that listen to the celestia-core network for blocks,
+validate and extend the block data in order to generate the DAH, and request both the ValidatorSet
+and Commit for that block height from the celestia-core network. Bridge nodes then package that data
+up into an ExtendedHeader and broadcast it to the headersub gossipsub topic. All nodes subscribed
+to headersub will receive and validate the ExtendedHeader, and store it, making it available to all
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply service/header/doc.go:13]: Here we define HeaderSub and use it everywhere below.

**Review by Wondertan — APPROVED** [review]
LGTM, nice work. 

Added some improvements

**Comment by YazzyYaz** [comment]
Hi @renaynay and @liamsi it looks good. Was thinking since it's a very high level documentation on running different types of node, maybe we can include somewhere "For more information on setting up a node and the hardware requirements needed, go visit our docs at https://docs.celestia.org" so we can centralize the information over there for all things docs related.

File: README.md (hunk 3 above)
Comment (renaynay) [reply README.md]: good point, resolved?

**Comment by renaynay** [comment]
> Hi @renaynay and @liamsi it looks good. Was thinking since it's a very high level documentation on running different types of node, maybe we can include somewhere "For more information on setting up a node and the hardware requirements needed, go visit our docs at https://docs.celestia.org" so we can centralize the information over there for all things docs related.

@YazzyYaz Yes this sounds good, do you mind making a suggestion on the PR that i can commit?

**Review by liamsi — APPROVED** [review]
Looks good to me. I think we can add a link to the central docs also in a followup PR :shipit:

**Pull request merged**